java.hints/src/org/netbeans/modules/java/hints/jdk/ConvertVarToExplicitType.java

```diff
@@ -0,0 +1,147 @@
+@Messages("MSG_ConvertibleToExplicitType=Convert var to explicit type")
+public class ConvertVarToExplicitType {
+
+    @TriggerPattern("$mods$ $type $var = $init") //NOI18N
+    public static ErrorDescription convertVarToExplicitType(HintContext ctx) {
+
+        if (!isLocalVarType(ctx)) {
+            return null;
+        }
+        TreePath treePath = ctx.getPath();
+
+        TreePath initTreePath = ctx.getVariables().get("$init");  //NOI18N
+        ExpressionTree t = ctx.getInfo().getTreeUtilities().parseExpression(initTreePath.getLeaf().toString(), null);
+        Scope s = ctx.getInfo().getTrees().getScope(ctx.getPath());
+        TypeMirror initTypeMirror = ctx.getInfo().getTreeUtilities().attributeTree(t, s);
+
+        TypeMirror variableTypeMiror = ctx.getInfo().getTrees().getElement(treePath).asType();
+
+        // variable initializer type should be same as variable type.
+        if ((variableTypeMiror.getKind() == TypeKind.ERROR) || (!ctx.getInfo().getTypes().isSameType(variableTypeMiror, initTypeMirror))) {
+            return null;
+        }
+        return ErrorDescriptionFactory.forTree(ctx, ctx.getPath(), Bundle.MSG_ConvertibleToExplicitType(),
+                new JavaFixImpl(ctx.getInfo(), ctx.getPath()).toEditorFix());
+    }
+
+    /**
+     * Fix for converting local 'var' type to explicit variable type
+     *
+     */
+    private static final class JavaFixImpl extends JavaFix {
+
+        public JavaFixImpl(CompilationInfo info, TreePath tp) {
+            super(info, tp);
+        }
+
+        @Override
+        @Messages("FIX_convertVarToExplicitType=Replace var with explicit type")
+        protected String getText() {
+            return Bundle.FIX_convertVarToExplicitType();
+        }
+
+        @Override
+        protected void performRewrite(TransformationContext tc) throws Exception {
+            WorkingCopy wc = tc.getWorkingCopy();
+            CompilationUnitTree cut = wc.getCompilationUnit();
+            TreePath statementPath = tc.getPath();
+
+            TreeMaker make = wc.getTreeMaker();
+
+            if (statementPath.getLeaf().getKind() == Tree.Kind.VARIABLE) {
+                VariableTree oldVariableTree = (VariableTree) statementPath.getLeaf();
+                TypeMirror type = wc.getTrees().getTypeMirror(statementPath);
+                VariableTree newVariableTree = make.Variable(
+                        oldVariableTree.getModifiers(),
+                        oldVariableTree.getName(),
+                        make.Type(type),
+                        oldVariableTree.getInitializer()
+                );
+                wc.rewrite(oldVariableTree, newVariableTree);
+            }
+        }
+
+    }
+
+    /**
+     *
+     * @param ctx : HintContext
+     * @return true if pre-conditions for hint to be enable is meet
+     */
+    private static boolean isLocalVarType(HintContext ctx) {
```

Comment:
Suggestion: this may be a good addition to Utilities in hints.errors (there are other type/tree related checks and manipulations). @jlahoda what do you think - shouldn't TreeUtilities check for lang version in their `isVarType` ?

Reply:
Moved version check to TreeUtilities.isVarType().

Reply:
Not sure about that - is there other methods in TU doing such checks?

It may be OK for var, as "var i = 0;" is not a var type under --source 9, but generally:
-imagine a hypothetical method checking if a tree is a lambda (e.g. TU.isLambda)
-running with --source 7 (which does not allow lambda), a lambda occurrence will cause an error. But the model may include the lambda due to error recovery. Should TU.isLambda return false for --source 7 for such error-recovered tree? I don't think it should - it is a lambda, just in a source level that does not support it. There may even be an error-fix that would convert the lambda to an anonymous class, and it should be able to use the TU.isLambda method, right?